File: dev/src/samples/NetSamplesVB/Samples/CogoPoints/UserDefinedPropertiesCommands.vb
Imports System.Collections.Generic

Imports Autodesk.AutoCAD.DatabaseServices
Imports Autodesk.AutoCAD.Runtime
Imports Autodesk.Civil.DatabaseServices

<Assembly: CommandClass(GetType(Autodesk.CivilizedDevelopment.UserDefinedPropertiesCommands))> 

Namespace Autodesk.CivilizedDevelopment
  Public Class UserDefinedPropertiesCommands
    Inherits SimpleDrawingCommand
    <CommandMethod("CDS_UDPDemo")> _
    Public Sub CDS_UDPDemo()
      createDemoPointsRepresentingWells()
      createDemoProperties()
      createDemoPointGroup()
      setDemoPointProperties()
    End Sub

    Private Sub createDemoPointsRepresentingWells()
      Dim generator As New RandomCoordinateGenerator()
      _civildoc.CogoPoints.Add(generator.GetCoordinates(10), "WELL")
    End Sub

    Private Sub createDemoProperties()
      _classification = createClassification("CivilizedDevelopment")
      addUDPs(_classification)
    End Sub

    Private Sub createDemoPointGroup()

      Dim id As ObjectId = _civildoc.PointGroups.Add("CivilizedDevelopment")
      Using tr As Transaction = startTransaction()
        customizePointGroup(id)
        tr.Commit()
      End Using

    End Sub

    Private Sub setDemoPointProperties()
      Using tr As Transaction = startTransaction()
        For Each pointId As ObjectId In _civildoc.CogoPoints
          customizePoint(pointId)
        Next

        tr.Commit()
      End Using
    End Sub

    Private Sub customizePointGroup(id As ObjectId)
      Dim group As PointGroup = TryCast(id.GetObject(OpenMode.ForWrite), 
        PointGroup)
      Dim query As New CustomPointGroupQuery()
      query.QueryString = "RawDescription='WELL*'"
      group.SetQuery(query)
      group.UseCustomClassification(_classification)
    End Sub

    Private Sub customizePoint(id As ObjectId)
      Dim point As CogoPoint = TryCast(id.GetObject(OpenMode.ForWrite), 
        CogoPoint)
      point.SetUDPValue(_potable, True)
      point.SetUDPValue(_chloride, 150)
      point.SetUDPValue(_laboratory, "CivilDev Labs, Inc.")
      point.SetUDPValue(_hardness, "Carbonate (Temporary)")
      point.SetUDPValue(_alkalinity, 7.5)
    End Sub

    Private Function createClassification(name As String) As UDPClassification
      Dim classifications As UDPClassificationCollection =
        UDPClassificationCollection.GetPointUDPClassifications(_database)
      Return classifications.Add(name)
    End Function

    Private Sub addUDPs(classification As UDPClassification)
      _potable = classification.CreateUDP(
        makeBoolDefinition("Potable",
          "Indicates if water is suitable for consumption."))
      _chloride = classification.CreateUDP(
        makeIntDefinition("Chloride", "Amount of chloride (mg/l)."))
      _laboratory = classification.CreateUDP(
        makeStringDefinition("Laboratory",
          "Name of the laboratory that performed analysis."))
      _hardness = classification.CreateUDP(
        makeEnumDefinition("Hardness",
          "Hardness with respect to anions in metallic cations."))
      _alkalinity = classification.CreateUDP(
        makePlainDoubleDefinition("Alkalinity",
          "Ability to resist sudden changes in pH."))
    End Sub

    Private Function makeBoolDefinition(name As String, description As String) _
        As AttributeTypeInfoBool

      Dim definition As New AttributeTypeInfoBool(name)
      definition.Description = description
      definition.DefaultValue = False
      definition.UseDefaultValue = True
      Return definition
    End Function

    Private Function makeIntDefinition(name As String, description As String) _
        As AttributeTypeInfoInt

      Dim definition As New AttributeTypeInfoInt(name)
      definition.Description = description
      definition.DefaultValue = 0
      definition.LowerBoundValue = Int32.MinValue
      definition.LowerBoundInclusive = True
      definition.UpperBoundValue = Int32.MaxValue
      definition.UpperBoundInclusive = True
      definition.UseDefaultValue = True
      Return definition
    End Function

    Private Function makeStringDefinition(name As String,
        description As String) As AttributeTypeInfoString

      Dim definition As New AttributeTypeInfoString(name)
      definition.Description = description
      definition.DefaultValue = [String].Empty
      definition.UseDefaultValue = True
      definition.Description = description
      Return definition
    End Function

    Private Function makeEnumDefinition(name As String,
        description As String) As AttributeTypeInfoEnum

      Dim valueDefinitions As New EnumDefinition()
      Dim definition As New AttributeTypeInfoEnum(name, valueDefinitions.Values)
      definition.Description = description
      definition.DefaultValue = valueDefinitions.DefaultValue
      definition.UseDefaultValue = True
      Return definition
    End Function

    Private Function makePlainDoubleDefinition(name As String,
        description As String) As AttributeTypeInfoDouble

      Dim definition As New AttributeTypeInfoDouble(name)
      definition.Description = description
      definition.DefaultValue = 0.0
      definition.UseDefaultValue = True
      definition.LowerBoundValue = 0.0
      definition.LowerBoundInclusive = True
      definition.UpperBoundValue = [Double].MaxValue
      definition.UpperBoundInclusive = True
      definition.UseDefaultValue = True
      Return definition
    End Function

    Private _classification As UDPClassification
    Private _potable As UDPBoolean
    Private _chloride As UDPInteger
    Private _laboratory As UDPString
    Private _hardness As UDPEnumeration
    Private _alkalinity As UDPDouble

    Private Class EnumDefinition
      Public Sub New()
        _values = New List(Of String)()
        _values.Add("Carbonate (Temporary)")
        _values.Add("Non-Carbonate (Permanent)")
      End Sub

      Public ReadOnly Property Values() As IEnumerable(Of String)
        Get
          Return _values
        End Get
      End Property

      Public ReadOnly Property DefaultValue() As String
        Get
          Return _values(0)
        End Get
      End Property

      Private _values As List(Of String)
    End Class
  End Class
End Namespace

File: dev/src/samples/NetSamples/Samples/CogoPoints/UserDefinedPropertiesCommands.cs
using System;
using System.Collections.Generic;

using Autodesk.AutoCAD.DatabaseServices;
using Autodesk.AutoCAD.Runtime;
using Autodesk.Civil.DatabaseServices;

[assembly: CommandClass(
    typeof(Autodesk.CivilizedDevelopment.UserDefinedPropertiesCommands))]

namespace Autodesk.CivilizedDevelopment
{
    public class UserDefinedPropertiesCommands : SimpleDrawingCommand
    {
        [CommandMethod("CDS_UDPDemo")]
        public void CDS_UDPDemo()
        {
            createDemoPointsRepresentingWells();
            createDemoProperties();
            createDemoPointGroup();
            setDemoPointProperties();
        }

        private void createDemoPointsRepresentingWells()
        {
            RandomCoordinateGenerator generator = 
                new RandomCoordinateGenerator();
            _civildoc.CogoPoints.Add(generator.GetCoordinates(10), "WELL");
        }

        private void createDemoProperties()
        {
            _classification = createClassification("CivilizedDevelopment");
            addUDPs(_classification);
        }

        private void createDemoPointGroup()
        {
            
            ObjectId id = _civildoc.PointGroups.Add("CivilizedDevelopment");
            using (Transaction tr = startTransaction())
            {
                customizePointGroup(id);
                tr.Commit();
            }
            
        }

        private void setDemoPointProperties()
        {
            using (Transaction tr = startTransaction())
            {
                foreach (ObjectId pointId in _civildoc.CogoPoints)
                {
                    customizePoint(pointId);
                }

                tr.Commit();
            }
        }

        private void customizePointGroup(ObjectId id)
        {
            PointGroup group = id.GetObject(OpenMode.ForWrite) as PointGroup;
            CustomPointGroupQuery query = new CustomPointGroupQuery();
            query.QueryString = "RawDescription='WELL*'";
            group.SetQuery(query);
            group.UseCustomClassification(_classification);
        }

        private void customizePoint(ObjectId id)
        {
            CogoPoint point = id.GetObject(OpenMode.ForWrite) as CogoPoint;
            point.SetUDPValue(_potable, true);
            point.SetUDPValue(_chloride, 150);
            point.SetUDPValue(_laboratory, "CivilDev Labs, Inc.");
            point.SetUDPValue(_hardness, "Carbonate (Temporary)");
            point.SetUDPValue(_alkalinity, 7.5);
        }

        private UDPClassification createClassification(string name)
        {
            UDPClassificationCollection classifications =
                UDPClassificationCollection
                .GetPointUDPClassifications(_database);
            return classifications.Add(name);
        }

        private void addUDPs(UDPClassification classification)
        {
            _potable = classification.CreateUDP(makeBoolDefinition("Potable", 
                "Indicates if water is suitable for consumption."));
            _chloride = classification.CreateUDP(makeIntDefinition("Chloride", 
                "Amount of chloride (mg/l)."));
            _laboratory = classification.CreateUDP(
                makeStringDefinition("Laboratory", 
                "Name of the laboratory that performed analysis."));
            _hardness = classification.CreateUDP(makeEnumDefinition("Hardness", 
                "Hardness with respect to anions in metallic cations."));
            _alkalinity = classification.CreateUDP(
                makeDoubleDefinition("Alkalinity", 
                "Ability to resist sudden changes in pH."));
        }

        private AttributeTypeInfoBool makeBoolDefinition(
            string name, string description)
        {
            AttributeTypeInfoBool definition = new AttributeTypeInfoBool(name);
            definition.Description = description;
            definition.DefaultValue = false;
            definition.UseDefaultValue = true;
            return definition;
        }

        private AttributeTypeInfoInt makeIntDefinition(
            string name, string description)
        {
            AttributeTypeInfoInt definition = new AttributeTypeInfoInt(name);
            definition.Description = description;
            definition.DefaultValue = 0;
            definition.LowerBoundValue = Int32.MinValue;
            definition.LowerBoundInclusive = true;
            definition.UpperBoundValue = Int32.MaxValue;
            definition.UpperBoundInclusive = true;
            definition.UseDefaultValue = true;
            return definition;
        }

        private AttributeTypeInfoString makeStringDefinition(
            string name, string description)
        {
            AttributeTypeInfoString definition = 
                new AttributeTypeInfoString(name);
            definition.Description = description;
            definition.DefaultValue = String.Empty;
            definition.UseDefaultValue = true;
            return definition;
        }

        private AttributeTypeInfoEnum makeEnumDefinition(
            string name, string description)
        {
            EnumDefinition valueDefinitions = new EnumDefinition();
            AttributeTypeInfoEnum definition = 
                new AttributeTypeInfoEnum(name, valueDefinitions.Values);
            definition.Description = description;
            definition.DefaultValue = valueDefinitions.DefaultValue;
            definition.UseDefaultValue = true;
            return definition;
        }

        private AttributeTypeInfoDouble makeDoubleDefinition(
            string name, string description)
        {
            AttributeTypeInfoDouble definition =
                new AttributeTypeInfoDouble(name);
            definition.Description = description;
            definition.DefaultValue = 0.0;
            definition.UseDefaultValue = true;
            definition.LowerBoundValue = 0.0;
            definition.LowerBoundInclusive = true;
            definition.UpperBoundValue = Double.MaxValue;
            definition.UpperBoundInclusive = true;
            definition.UseDefaultValue = true;
            definition.DataType = AttributeTypeInfoDoubleDataType.Percent;
            return definition;
        }

        private UDPClassification _classification;
        private UDPBoolean _potable;
        private UDPInteger _chloride;
        private UDPString _laboratory;
        private UDPEnumeration _hardness;
        private UDPDouble _alkalinity;

        class EnumDefinition
        {
            public EnumDefinition()
            {
                _values = new List<string>();
                _values.Add("Carbonate (Temporary)");
                _values.Add("Non-Carbonate (Permanent)");
            }

            public IEnumerable <string> Values
            {
                get
                {
                    return _values;
                }
            }

            public string DefaultValue
            {
                get
                {
                    return _values[0];
                }
            }

            private List<string> _values;
        }
    }
}